### PR TITLE
fix(tabs): let an agent's own /rename win the tab label

### PR DESCRIPTION
**Pull request by lafraia**
## Problem

Claude's `/rename` only reached the tab label when it was the first thing you did in a session. After any prompt, the tab kept a name derived from that opening message and every later `/rename` was ignored.

The cause is that Claude publishes both a deliberate `/rename` and its auto-generated task summaries on the same OSC title channel. Orca could not tell them apart, so `tab-title-resolution.ts` ranked Orca's first-prompt `generatedTitle` above the live title — the right call for summaries, wrong for renames.

## Approach

The transcript does separate them. `/rename` appends a `custom-title` record; generated summaries append `ai-title`:

```jsonc
{"type": "custom-title", "customTitle": "billing-fix", "sessionId": "…"}   // deliberate rename
{"type": "ai-title",     "aiTitle": "Fix the intake flow"}                 // generated summary
```

So the rename gets its own tab field (`agentSessionTitle`), fed by watching the transcript the hook feed already names per pane, and kept apart from the live title.

`/rename` is a local slash command that fires no hook, which shapes the design: the watch is driven by the transcript file itself, and reconciled against the full provider-session set on a timer. The hook change feed never observes a rename, and it reports only *changed* identities — reconciling against that delta would tear down every other pane's watch. A session restored at startup would likewise never get a watch without the timer.

## Precedence

A manual tab rename and the agent's `/rename` are both the user renaming, so neither outranks the other by origin. Each carries a timestamp and the newer act wins; ties go to the manual rename.

```
newest rename → quick command → OpenCode session → unconfirmed live name → generated → live title
```

The "unconfirmed live name" rung is new too: an OSC title that carries a real name, as opposed to a status frame, identity echo (`✳ Claude Code`), spinner, or cwd. It covers agents whose renames Orca cannot confirm against a transcript.

## Testing

Verified end-to-end in the running app by reading the painted `data-tab-title` over CDP, in both directions: `/rename` after a manual rename flips the label, and a manual rename after `/rename` flips it back.

New coverage sits at each layer, including the projection between the store and the tab strip. Three layers rebuild the terminal tab field-by-field, and a dropped stamp there silently reinstates the old manual-rename-always-wins order without failing any resolver test.

Full suite green (41883 tests), plus typecheck, oxlint native + type-aware, oxfmt, and the max-lines ratchet.

## Not covered

SSH panes. Their transcript lives on the remote host, so the read fails and the watch degrades quietly — those tabs keep following the live OSC title. Reading the transcript over the relay is separate work.

`getActivityThreadTaskTitle` is left alone on purpose: it resolves a per-turn *task* title, not the tab name.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/main/agent-hooks/claude-session-rename-watch.ts [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        const existing = entries.get(source.paneKey)
+        if (existing) {
+          if (existing.transcriptPath === transcriptPath) {
+            // Why: fs.watch binding is best-effort on network filesystems, so
+            // retry on every reconcile rather than trusting the first bind.
+            if (existing.watcher.needsRebind()) {
+              existing.watcher.bind()
+            }
+            continue
+          }
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/main/agent-hooks/claude-session-rename-watch.ts:184]: Applied in ceca1509f. `sync()` now re-reads when a late `bind()` succeeds:

```ts
if (existing.watcher.needsRebind() && existing.watcher.bind()) {
  // Why: a rename written while the watcher was unbound produced no
  // event, so catch up now — the next write may never come.
  read(source.paneKey, existing)
}
```

File: src/shared/claude-session-custom-title.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  // Why: the first line of a tail read can be a fragment of a longer record;
+  // JSON.parse rejects it, so no separate boundary handling is needed.
+  for (let index = lines.length - 1; index >= 0; index -= 1) {
+    const title = extractClaudeCustomTitle(lines[index] ?? '')
+    if (title) {
+      return title
+    }
+  }
+  return null
+}
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/shared/claude-session-custom-title.ts]: Confirmed, and worse than a mismatch with the docstring — the docstring described the intended behaviour and the code did not implement it.

Applied in ceca1509f, taking the split you proposed. `parseClaudeCustomTitleLine` returns `''` for a cleared rename and `null` only when the line is not a `custom-title` record, so the backward scan stops at the newest matching record either way.

I also added the ordering test you offered: `[A, ""]` resolves to null, and `[A, "", B]` resolves to `B`.

File: src/renderer/src/components/tab-group/useTabGroupWorkspaceModel.ts [hunk 3]
```diff
@@ -131,7 +131,10 @@ export function useTabGroupWorkspaceModel({
               {
                 ...item,
                 quickCommandLabel: item.quickCommandLabel ?? terminalTab?.quickCommandLabel,
-                generatedLabel: item.generatedLabel ?? terminalTab?.generatedTitle
+                generatedLabel: item.generatedLabel ?? terminalTab?.generatedTitle,
+                customLabelAt: item.customLabelAt ?? terminalTab?.customTitleAt,
+                agentSessionLabel: item.agentSessionLabel ?? terminalTab?.agentSessionTitle,
+                agentSessionLabelAt: item.agentSessionLabelAt ?? terminalTab?.agentSessionTitleAt
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/renderer/src/components/tab-group/useTabGroupWorkspaceModel.ts:140]: Confirmed — a real bug, and the one that made this fail in the running app. With a locally renamed tab the unified `customLabel` is null, so passing `customTitleAt` without its title made the resolver read the manual rename as absent and hand the label to a *staler* agent rename.

Fixed in ceca1509f in both projections, plus a test with a null unified `customLabel` and a terminal-only `customTitle`.

File: src/main/index.ts [hunk 4]
```diff
@@ -2062,15 +2068,39 @@ void app.whenReady().then(async () => {
...
+  // Why: the change feed only fires on a hook, so a session restored at startup
+  // (or one whose only activity is `/rename`) would never get a watch. Reconcile
+  // on a timer too — unchanged panes just re-check their fs.watch binding, which
+  // the native watcher documents as best-effort and caller-owned.
+  syncClaudeSessionRenameWatch()
+  const claudeSessionRenameReconcileTimer = setInterval(
+    syncClaudeSessionRenameWatch,
+    CLAUDE_SESSION_RENAME_RECONCILE_MS
+  )
+  claudeSessionRenameReconcileTimer.unref?.()
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/main/index.ts:2106]: Took the snapshot option in ceca1509f — `agentSession:getRenameSnapshot`, mirroring `agentStatus:getSnapshot`, whose own comment describes this same startup race ("startup cannot lose replayed statuses while its local store is still empty"). The watch exposes `getKnownRenames()`; the renderer pulls once, right after subscribing.

Worth recording: the first version of that pull broke 9 unrelated tests in `useIpcEvents.test.ts` and `mobile-terminal-reveal-tab-adoption.test.ts`. Their `window.api` mock returns a listener stub for any unknown method, so the call returned a function and `.then` threw — taking down registration of every IPC listener in that mount, not just this one. Rather than patch the mocks, the pull is now a guarded async IIFE: a best-effort catch-up must never abort the rest of the hook.

File: src/shared/types.ts [hunk 5]
```diff
@@ -873,6 +878,17 @@ export type TerminalTab = {
...
+   *  (`agentSessionTitle`); both are the user renaming, so the newer act wins.
+   *  Absent on tabs renamed before this field existed — treated as oldest. */
+  customTitleAt?: number
+  /** The name the agent's own rename command set for this session (Claude
+   *  `/rename`, persisted as a `custom-title` transcript record). Kept apart
+   *  from `title` because the agent publishes renames and auto-generated task
+   *  summaries on the same OSC channel — only this one is a deliberate rename. */
+  agentSessionTitle?: string | null
+  /** When `agentSessionTitle` was observed, for the newer-rename-wins compare. */
+  agentSessionTitleAt?: number
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/shared/types.ts:891]: Applied in ceca1509f — `terminalTabEqual` now compares `customTitleAt` and `agentSessionTitleAt` alongside the titles they order.

File: src/renderer/src/store/pinned-tab-close-guard.test.ts [hunk 6]
```diff
@@ -140,7 +140,16 @@ describe('resolvePinnedTabLabel', () => {
...
+            // Why: a status frame carries no name, so the generated label wins.
+            label: '✳ Claude Code'
+          },
+          {
+            id: 'e',
+            entityId: 'ee',
+            customLabel: null,
+            quickCommandLabel: null,
+            generatedLabel: 'Gen',
+            label: ' Renamed session '
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/renderer/src/store/pinned-tab-close-guard.test.ts:155]: Right — the assertion reached its result through the live-label rung, so it never exercised the agent-rename one. `agentSessionLabel` added in ceca1509f.

File: src/renderer/src/store/slices/agent-session-rename-tab-title.test.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    expect(terminalTab(store).agentSessionTitle).toBeUndefined()
+
+    store.getState().setTabAgentSessionTitle(tabId, 'billing-fix')
+    const stampedAt = terminalTab(store).agentSessionTitleAt
+    expect(stampedAt).toBeTypeOf('number')
+
+    // Why: Claude re-appends the rename record every turn; restamping it would
+    // keep pushing a stale rename past a newer manual one.
+    store.getState().setTabAgentSessionTitle(tabId, 'billing-fix')
+    expect(terminalTab(store).agentSessionTitleAt).toBe(stampedAt)
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/renderer/src/store/slices/agent-session-rename-tab-title.test.ts:67]: Applied in ceca1509f: fake timers, with `advanceTimersByTime` between the two calls so a restamping regression cannot pass by landing in the same millisecond.

File: src/main/agent-hooks/claude-session-rename-watch.ts [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    },
+    getKnownRenames(): ClaudeSessionRename[] {
+      const renames: ClaudeSessionRename[] = []
+      for (const [paneKey, entry] of entries) {
+        if (entry.lastTitle) {
+          renames.push({ paneKey, customTitle: entry.lastTitle })
+        }
+      }
+      return renames
+    },
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/main/agent-hooks/claude-session-rename-watch.ts:221]: Confirmed and fixed in 5c3395b8a — an empty read now forgets the cached title, so `getKnownRenames()` stops replaying a name the transcript has dropped. Covered by a clear-then-snapshot test.

I did not take the second half — propagating an explicit clear through the rename IPC contract. Clearing the session name inside Claude is not a request to un-name the Orca tab. The model here is "the newer rename wins", and a clear is not a rename to something else, so the tab keeps whatever it is showing until some rename supersedes it. That keeps the stale-replay bug fixed without inventing clear semantics for tabs, which would be a product decision rather than a correctness one. Happy to revisit if the maintainers read it the other way.

File: src/renderer/src/hooks/useIpcEvents.ts [hunk 9]
```diff
@@ -3317,6 +3318,34 @@ export function useIpcEvents(): void {
...
+        if (!Array.isArray(renames)) {
+          return
+        }
+        for (const rename of renames) {
+          applyAgentSessionRename(rename)
+        }
+      } catch {
+        // Why: a best-effort catch-up. Registering the rest of this hook's IPC
+        // listeners matters more than recovering a startup rename.
+      }
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/renderer/src/hooks/useIpcEvents.ts:3358]: Confirmed — the race is real, and self-inflicted: the listener is live before the pull is answered, so a mid-flight rename was overwritten by the older snapshot value and restamped newer.

Fixed in 5c3395b8a, though not with a revision on the payload. The snapshot is older than any live event by construction, so it only ever needs to fill a gap: `setTabAgentSessionTitle` gained `onlyWhenUnset`, which the snapshot path passes and the live path does not. A response landing after the effect is torn down is dropped outright. Two tests cover it — snapshot loses to a rename that already landed, snapshot applies when nothing did.

That felt proportionate to a pull whose only job is covering the startup gap; a monotonic revision would be the right call if the snapshot ever became a general reconciliation channel.

File: src/main/agent-hooks/claude-session-rename-watch.test.ts [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+    // Why: with the watcher unbound, a rename produces no event at all. Only the
+    // catch-up read on a successful rebind can surface it.
+    const stub = stubs.forPath(transcriptPath)
+    if (stub) {
+      stub.rebindNeeded = true
+    }
+    await writeFile(transcriptPath, renameRecord('written-while-unbound'))
+    watch?.sync([{ paneKey: 'tab-1:leaf-1', transcriptPath }])
+
+    await waitFor(() => renames.length === 2, 'the catch-up read after rebind')
```
[the remark this answers is not in the document]
Comment (lafraia) [reply src/main/agent-hooks/claude-session-rename-watch.test.ts:225]: Right — I added the `bindFails` switch and then never flipped it, so the test only proved the happy path. Fixed in c2135f40c, driving both halves: a reconcile whose bind fails emits nothing, and only the reconcile that binds successfully performs the catch-up read.

Writing it caught the stub lying about the failure. It left `bound` untouched on a failed `bind()`, while the real watcher stays detached when `watch()` throws — so the assertion I meant to add was checking against a fiction. The stub now models the detach.

One extra assertion beyond the suggestion: `bindAttempts` is captured before the failing reconcile and asserted to have grown. Without it, "no rename emitted" would hold just as well if no bind had been attempted at all, and the silent half would pass vacuously.

**Comment by nwparker** [comment]
Thanks for the report and the PR @lafraia — I reproduced this independently and landed a fix in #12481 that reaches the same root cause (`custom-title` vs `ai-title` in the transcript). Two things from your diff were genuinely better than what I had, and I took them, with credit:

- **The tab-strip rebuild.** Your point that `useTabGroupWorkspaceModel` / `FloatingTerminalPanel` rebuild the terminal tab field-by-field is real and easy to miss — my branch had exactly that bug: the store, the unified tab and both resolvers all had the rename while the painted `data-tab-title` still showed the generated title, with every resolver test green. Fixed, with an e2e that reads the painted label.
- **The bounded tail read.** I replaced a full-file scan with your 256 KB tail. One caveat from measuring real transcripts: re-appending is common (a 2.8 MB session had 37 records, newest 403 B from EOF) but not guaranteed — one session had a single record at the top and never re-appended, so I widen to a full scan when the tail holds no record.

One thing worth a look in your version: `resolveNewestTabRename` orders the manual rename against the agent rename by timestamp, and `customTitleAt` is a new field. A tab renamed before the upgrade has no stamp, so it reads as oldest — on the first launch after the change, the startup reconcile finds any pre-existing `custom-title` in that pane's transcript, stamps it with `Date.now()`, and it outranks the user's own tab name. Might be worth treating a missing `customTitleAt` as newer than an agent rename observed after startup, or seeding the stamp on migration.

Also flagging a deliberate difference rather than a defect: your "unconfirmed live name" rung means Claude's `ai-title` summaries outrank Orca's `generatedTitle` too (your `agent-row-conversation-name` expectation flip captures this). That's a reasonable product call, but it's broader than this bug, so I left it out and kept "auto summaries lose" intact.

**Comment by nwparker** [comment]
@lafraia If you want this fixed, you can open an issue and collect upvotes.
See https://github.com/stablyai/orca/pull/12481#issuecomment-5229939649

The complexity IMO is too great right now. Sadly not a simple fix

**Comment by Jinwoo-H** [comment]
Current main lacks the agent-owned /rename tab-label watcher. We are taking the ownership fix into owner implementation under #17539, alongside related issue #15497. Closing the submitted branch; full credit to @lafraia. No action is requested from you. Thank you.\n\nOwner issue: https://github.com/stablyai/orca/issues/17539
